source/plasma3DClass.py

```diff
@@ -0,0 +1,1573 @@
+			f.write('pi=\t3.141592653589793\n')
+			f.write('2*pi=\t6.283185307179586\n')
+
+
+	def writeM3DC1supFile(self):
+		"""
+		Write M3D-C1 supplemental input file
+		Overwrites any existing one.
+		"""
+		with open(self.cwd + '/' + 'm3dc1sup.in', 'w') as f:
+			for i in range(len(self.C1Files)):
+				f.write(self.C1Files[i] + '\t' + str(self.C1scales[i]) + '\t' + str(self.C1phases[i]) + '\n')
+
+
+	def writeCoilsupFile(self, machine = None):
+		"""
+		This would be machine specific and needs updating in the future
+		"""
+		# with open(self.cwd + '/' + 'heatsup.in', 'w') as f:
+		#	pass
+
+		return
+		
+	
+	def wait2finish(self, NCPUs, tag):
+		import time
+		print ('Waiting for job to finish...', end='')
+		time.sleep(5)	# wait 5 seconds
+		while(self.isProcessRunning()):
+			time.sleep(60)		# wait 1 minute; no CPU usage
+		print('done')
+			
+		if not self.isComplete():
+			print('MAFOT run ended prematurely. Attempt restart...')
+			subprocess.call(['mpirun','-n',str(NCPUs),'heatlaminar_mpi','-P','points.dat','_lamCTL.dat',tag])
+			self.wait2finish(NCPUs, tag)
+		else: return
+	
+	
+	def isComplete(self, logsPath = None):
+		if logsPath is None: 
+			logsPath = self.cwd
+		if not logsPath[-1] == '/': logsPath += '/'
+
+		allFiles = os.listdir(logsPath)
+		fileList = [f for f in allFiles if '_Master.dat' in f]
+		for file in fileList:
+			lines = subprocess.check_output(['tail', file])
+			lines = lines.decode('UTF-8').split('\n')
+			for line in lines: 
+				if 'Program terminates normally' in line: 
+					return True
+			else: 
+				return False
+	
+	
+	def isProcessRunning(self):
+		import getpass
+		user = getpass.getuser()
+		lines = subprocess.check_output('ps aux | grep heatlaminar_mpi', shell = True)	# for some reason only works with shell = True
+		lines = lines.decode('UTF-8').split('\n')
+		for line in lines:
+			if ('mpirun' in line) & (user in line):
+				self.pid = int(line.strip().split()[1])
+				return True
+		self.pid = -1
+		return False
+		
+
+#==========================================================================================================================
+#   heatflux3D class
+#==========================================================================================================================
+class heatflux3D:
+	"""
+	This class gets the penetration depth and connection length of unshadowed points
+	and calculates the parallel heat flux
+	still needs normalization to power balance
+	still needs incident angle
+	Example call:
+	"""
+
+	def __init__(self):
+		self.psimin = None
+		self.Lc = None		# in km
+		self.N = 1
+		self.NCPUs = 100
+		self.q = np.zeros(self.N)
+		self.q0 = None
+		self.ep = None	# equilParams_class instance for EFIT equilibrium
+		self.HFS = None	# True: use high field side SOL, False: use low field side SOL
+		self.teProfileData = None
+		self.neProfileData = None
+		self.allowed_vars = ['Lcmin', 'lcfs', 'lqCN', 'S', 'P', 'radFrac', 'qBG', 
```

Comment:
@wingena see how I had to change Pinj to P and coreRadFrac to radFrac to keep this code consistent with the rest of the code base and input files.